### PR TITLE
Adds the "Royal Visit" Station trait that brings a feudal prince/princess to visit the station.

### PR DESCRIPTION
**Pull request by GearTinkster**
## About The Pull Request

This is an idea I came across in Fulp when some guy talked about it but nobody ever implemented it, so this is my attempt at it! Basically this PR adds a neutral station trait that adds an "MVP" of some kind. This MVP is **very likely** to be a target of antagonistic antics. This makes it important for security to keep a close eye on them to make sure they don't get taken out!

Although this PR specifically adds a royal feudal lord (think Rimworld-esque), this PR also adds a framework for more protagonist/MVP characters! So if we want a mafioso that has dark dealings with the station or a group of orphans that need protecting all of that is possible in future PRs!

For now though, its just the prince/princess. They have the following traits:

-  They come equipped with plate armor, a buckler, and a claymore. Good for fighting off evil!
- They fear the supernatural, may the gods not smite them.
- They cannot use guns, they are magical weapons beyond their comprehension
- They cannot wear badbacks. Their backs are frail from the luxuries of their life.
- They look great, expect them to really make any room look good.
- They speak in an ancient accent, only god knows what half of the words they say are.

I would expect players to take the initiative in inventing how to play this role, for all intents and purposes they are expected to side with the station and work with them to survive, all while roleplaying the role of an ambassador of their house. 

## Why It's Good For The Game

Something like this can put a twist on the normal cloak and dagger gameplay of SS13 by revealing one crucial detail: Who the potential target might be. This gives security an active idea of who to protect but also can open up interesting challenges for people who have played a lot of antagonist roles before.

## Changelog
:cl: GearTinkster
expansion: Adds a new station trait: Royal Visit, which will let a member of a local royal family visit the station. Sure to be a target for any of the local factions!
qol: Changed the font color for the greenannounce span as it was a neon green. Very ugly! Not cool! Hurt my eyes :(
/:cl:

### CODE REVIEW

**Comment by GearTinkster** [comment]
I hope the code is okay by the way! this was a difficult one to figure out especially with how antagonist spawning works! :)

File: code/__DEFINES/role_preferences.dm [hunk 1]
```diff
@@ -42,7 +42,9 @@
 #define ROLE_FAMILIES "Familes Antagonists"
 #define ROLE_SYNDICATE_CYBERSUN "Cybersun Space Syndicate" //Ghost role syndi from Forgottenship ruin
 #define ROLE_SYNDICATE_CYBERSUN_CAPTAIN "Cybersun Space Syndicate Captain" //Forgottenship captain syndie
-#define ROLE_FREE_GOLEM "Free Golem" 
+#define ROLE_FREE_GOLEM "Free Golem"
+#define ROLE_PROTAGONIST "Protagonist"
```
Comment (tralezab) [inline code/__DEFINES/role_preferences.dm:46]: may as well doc these

File: code/__DEFINES/role_preferences.dm [hunk 2]
```diff
@@ -42,7 +42,9 @@
 #define ROLE_FAMILIES "Familes Antagonists"
 #define ROLE_SYNDICATE_CYBERSUN "Cybersun Space Syndicate" //Ghost role syndi from Forgottenship ruin
 #define ROLE_SYNDICATE_CYBERSUN_CAPTAIN "Cybersun Space Syndicate Captain" //Forgottenship captain syndie
```
Comment (tralezab) [inline code/__DEFINES/role_preferences.dm:44]: and while you're here if you could add a slash to these and put them above the defines so they count as docs, that'd be fantastic

File: code/datums/mutations/passive.dm [hunk 3]
```diff
@@ -11,3 +11,26 @@
 /datum/mutation/human/biotechcompat/on_losing(mob/living/carbon/human/owner)
 	owner.adjust_skillchip_complexity_modifier(-1)
 	return ..()
+
+/datum/mutation/human/good_looking
```
Comment (tralezab) [inline code/datums/mutations/passive.dm:15]: just fyi, this means anyone can find this mutation naturally. if you don't want that, take a look at how the holy mutations (honorbound, burdened) work to not be in the pool

File: code/datums/mutations/speech.dm [hunk 4]
```diff
@@ -207,6 +207,7 @@
 	quality = MINOR_NEGATIVE
 	text_gain_indication = "<span class='notice'>You feel like seeking the holy grail!</span>"
 	text_lose_indication = "<span class='notice'>You no longer feel like seeking anything.</span>"
+	var/has_startings = TRUE
```
Comment (tralezab) [inline code/datums/mutations/speech.dm:210]: document this var please! 📝

File: code/datums/mutations/speech.dm [hunk 5]
```diff
@@ -236,7 +237,12 @@
...
-		var/chosen_starting = pick(startings)
-		message = "[chosen_starting] [message]"
+		if(has_startings)
+			var/chosen_starting = pick(startings)
+			message = "[chosen_starting] [message]"
 
 		speech_args[SPEECH_MESSAGE] = message
+
+
+/datum/mutation/human/medieval/no_startings
```
Comment (tralezab) [inline code/datums/mutations/speech.dm:247]: this will also be in the random roll pool and it definitely shouldn't (see my other comment on this)

File: code/datums/station_traits/neutral_traits.dm [hunk 6]
```diff
@@ -75,3 +75,77 @@
...
+
+/datum/station_trait/protagonist/New()
+	. = ..()
+	RegisterSignal(SSdcs, COMSIG_GLOB_PRE_GAMEMODE_SETUP, .proc/on_gamemode_setup)
+	RegisterSignal(SSdcs, COMSIG_GLOB_POST_GAMEMODE_SETUP, .proc/after_gamemode_setup)
+	antag_datum_instance = new role_to_give() //Create this early for things such as family name to be setup
+
+/// Checks if candidates are connected and if they are banned or don't want to be the antagonist.
+/datum/station_trait/protagonist/proc/trim_candidates(list/candidates)
+	for(var/mob/dead/new_player/candidate_player in candidates)
```
Comment (tralezab) [inline code/datums/station_traits/neutral_traits.dm:103]:
```suggestion
	for(var/mob/dead/new_player/candidate_player as anything in candidates)
```
assuming that list/candidates is a list with only new players in it, adding "as anything" makes byond not try and filter the list for only candidate_player's type

File: code/datums/station_traits/neutral_traits.dm [hunk 7]
```diff
@@ -75,3 +75,77 @@
...
+			continue
+
+		if(!(ROLE_PROTAGONIST in candidate_client.prefs.be_special) || is_banned_from(candidate_player.ckey, list(ROLE_PROTAGONIST, ROLE_SYNDICATE)))
+			candidates.Remove(candidate_player)
+			continue
+
+/datum/station_trait/protagonist/proc/on_gamemode_setup(datum/gamemode/gamemode_ref)
+	SIGNAL_HANDLER
+	var/list/candidates = list()
+	for(var/i in GLOB.new_player_list)
```
Comment (tralezab) [inline code/datums/station_traits/neutral_traits.dm:120]: instead of
``` dm
for(var/i in GLOB.new_player_list)
    var/mob/dead/new_player/player = i
```
you should use
``` dm
for(var/mob/dead/new_player/player as anything in GLOB.new_player_list)
```
less boilerplate, and it avoids the use of var/i which we want more descriptive var names

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#avoid-unnecessary-type-checks-and-obscuring-nulls-in-lists
and
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/datums/station_traits/neutral_traits.dm [hunk 8]
```diff
@@ -75,3 +75,77 @@
...
+	name = "Royal Visit"
+	report_message = "This station has received an esteemed guest! They will most likely be a high value target for any Syndicate invasions. Be sure to keep them safe! It seems this guest is a royal prince or princess! Although they are of royal blood, they don't seem to be the most brave person."
+	role_to_give = /datum/antagonist/protagonist/scaredy_prince
+	weight = 4
+	trait_flags = NONE
+
+///Pick a family name and put it in the title!
+/datum/station_trait/protagonist/scaredy_prince/New()
+	. = ..()
+	if(istype(antag_datum_instance, /datum/antagonist/protagonist/scaredy_prince))
```
Comment (tralezab) [inline code/datums/station_traits/neutral_traits.dm:149]: is there a case where `antag_datum_instance` would not be `/datum/antagonist/protagonist/scaredy_prince` if this station trait is a subtype? are you trying to check if `antag_datum_instance` exists instead?

File: code/game/gamemodes/objective.dm [hunk 9]
```diff
@@ -116,37 +116,40 @@ GLOBAL_LIST_EMPTY(objectives)
...
-		if(!possible_targets.len)
-			possible_targets = all_possible_targets
-	if(possible_targets.len > 0)
-		target = pick(possible_targets)
+	if(SSstation.protagonists.len && prob(80))
+		target = pick(SSstation.protagonists)
+	else
+		var/list/possible_targets = list()
+		var/try_target_late_joiners = FALSE
+		for(var/I in owners)
```
Comment (tralezab) [inline code/game/gamemodes/objective.dm:124]: extra credit, if you want: this is all old objective code, so if you could apply the "as anything" to for loops iterating through types it knows and also apply the "descriptive variables" standard of naming variables, that'd be sick

File: code/modules/antagonists/protagonists/protagonist.dm [hunk 10]
```diff
@@ -0,0 +1,39 @@
+/datum/antagonist/protagonist
+	name = "Protagonist"
+	roundend_category = "protagonists"
+	antagpanel_category = "Protagonists"
+	job_rank = ROLE_PROTAGONIST
+	show_to_ghosts = TRUE
+	///Some protagonists have min ages
+	var/min_age = 16
```
Comment (tralezab) [inline code/modules/antagonists/protagonists/protagonist.dm:8]: our minimum age is 17, so prepare for intense heckling if you keep this

File: code/modules/antagonists/protagonists/protagonist.dm [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+	///Some protagonists have max ages
+	var/max_age = 99
+	///Outfit to put onto the character
+	var/outfit_type = /datum/outfit/wizard
```
Comment (tralezab) [inline code/modules/antagonists/protagonists/protagonist.dm:12]: make this the naked outfit on purpose to signal someone forgetting to set this? probably not needed

File: code/modules/antagonists/protagonists/protagonist.dm [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+/datum/antagonist/protagonist/on_gain()
+	. = ..()
+	equip_protagonist()
+	create_objectives()
+
+/datum/antagonist/protagonist/proc/equip_protagonist()
+	if(!owner)
+		CRASH("Antag datum with no owner.")
+	var/mob/living/carbon/human/H = owner.current
```
Comment (tralezab) [inline code/modules/antagonists/protagonists/protagonist.dm:23]: better var names, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/modules/antagonists/revolution/revolution.dm [hunk 13]
```diff
@@ -303,6 +303,18 @@
 		new_target.target = M
 		new_target.update_explanation_text()
 		objectives += new_target
+
+	var/untracked_protagonists = SSstation.protagonists
+	for(var/datum/objective/mutiny/O in objectives)
```
Comment (tralezab) [inline code/modules/antagonists/revolution/revolution.dm:308]: better var names, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/modules/antagonists/revolution/revolution.dm [hunk 14]
```diff
@@ -303,6 +303,18 @@
...
+		untracked_protagonists -= O.target
+
+	for(var/datum/mind/M in untracked_protagonists)
```
Comment (tralezab) [inline code/modules/antagonists/revolution/revolution.dm:311]: better var names, please. also, use "as anything" here if `untracked_protagonists` is only minds

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names
and
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#avoid-unnecessary-type-checks-and-obscuring-nulls-in-lists

File: code/datums/mutations/passive.dm [hunk 15]
```diff
@@ -11,3 +11,26 @@
...
+	name = "Good Looking"
+	desc = "A long lineage of strong genes has made this person look somewhat good compared to most, although it does make them a bit vain most likely."
+	quality = POSITIVE
+	text_gain_indication = "<span class='notice'>You feel like you're looking good!</span>"
+	text_lose_indication = "<span class='notice'>You no longer feel like like a prince charming.</span>"
+
+/datum/mutation/human/good_looking/on_acquiring(mob/living/carbon/human/owner)
+	if(..())
+		return
+	owner.AddElement(/datum/element/beauty, 2500) //what a sigh to behold!
```
Comment (tralezab) [inline code/datums/mutations/passive.dm:25]: most likely optional but is there any good way to make `2500` a define? i guess it's out of scope but there really is no reference for what values of beauty there are

**Review by tralezab — COMMENTED** [review]
this code's not too bad! i did a review for this repo's code standards, but maintainers may have more to say.

File: code/datums/mutations/passive.dm (hunk 3 above)
Comment (GearTinkster) [reply code/datums/mutations/passive.dm:15]: yeah looking good is reserved for Habsburgs

File: interface/stylesheet.dm [hunk 16]
```diff
@@ -74,7 +74,7 @@ h1.alert, h2.alert		{color: #000000;}
 .announce				{color: #228b22;	font-weight: bold;}
 .boldannounce			{color: #ff0000;	font-weight: bold;}
 .minorannounce			{					font-weight: bold;  font-size: 3;}
-.greenannounce			{color: #00ff00;	font-weight: bold;}
+.greenannounce			{color: #008f00;	font-weight: bold;}
```
Comment (Mothblocks) [inline interface/stylesheet.dm:77]: Why this change? Did we use this for anything else?

Comment (GearTinkster) [reply interface/stylesheet.dm:77]: forgot to mention this sorry! this is used for awards (I think those are achievements?) but it is a very neon green at the moment which is unreadable!
image

Comment (Mothblocks) [reply interface/stylesheet.dm:77]: Fair enough--put it in the change log though. I don't really know what tag you'd use...but should be in there.

**Comment by EOBGames** [comment]
Hi, I'm the loremaster for TG: I hate to do this, but I'm not sure this works flavourwise for the universe. The general idea of an MVP visiting the station is good, and I think that should be something carried forward, but I take issue with the flavour: what basically amounts to a fantasy prince/princess showing up to the station is just a bridge too far in my eyes.

One suggestion I would have for alternate flavours would be the son/daughter of a Nanotrasen executive- just as important, but a bit more fitting with the universe. Other ideas could be station inspectors, celebrities, or something along those lines.

I'm not going to block the PR or anything over this, but I'd really prefer if you took this into consideration.

File: interface/stylesheet.dm (hunk 16 above)
Comment (optimumtact) [reply interface/stylesheet.dm:77]: make sure you check it in dark mode too

**Comment by MrMelbert** [comment]
Who gets selected for the role? A random player who's readied up? A ghost? Is there an antagonist preference for it?

**Comment by Mothblocks** [comment]
> Is there an antagonist preference for it? 

Yes.

File: code/datums/station_traits/neutral_traits.dm [hunk 17]
```diff
@@ -75,3 +75,77 @@
...
+		var/datum/antagonist/protagonist/scaredy_prince/prince_antag_datum = antag_datum_instance
+		name = "Royal prince of house [prince_antag_datum.family_name]"
```
Comment (Iamgoofball) [inline code/datums/station_traits/neutral_traits.dm:151]: Royal Prince -> Royalty for gender agnostic naming

**Comment by GuillaumePrata** [comment]
I love the idea but I agree with EOB here, mostly this part:

> what basically amounts to a fantasy prince/princess showing up to the station is just a bridge too far in my eyes.

Having knight armor is silly for royalty, some luxurious clothing would be more fitting and for self defense a ~~Captain's~~ Officer's Sabre would work better.

I also dislike them not being able to use guns, yes you can say it's magical weapons but they can just go and set up the Supermatter and build mechs, maybe the bad aim quirk would be nice tho.

**Comment by Iamgoofball** [comment]
@EOBGames I think we can work out that it's like, a minor feudal county or something from outside the SSC, therefore the SSC wants to maintain good diplomatic relations with them.

File: code/datums/station_traits/neutral_traits.dm (hunk 8 above)
Comment (GearTinkster) [reply code/datums/station_traits/neutral_traits.dm:149]: its just there for safety but I suppose it's impossible for the datum to be wrong :)

File: code/modules/antagonists/protagonists/protagonist.dm (hunk 10 above)
Comment (GearTinkster) [reply code/modules/antagonists/protagonists/protagonist.dm:8]: it shouldn't matter then! :) you can't go lower this is just a minimum specific to the role, but if you can't even get 16 in prefs then you can't get 16 here. Its more for protags that might need to be above 30 for example.

**Comment by LemonInTheDark** [comment]
> I fully agree with Potato here, and will be a little less diplomatic about it: the general idea is good, but the theme is not:

I am become loremaster, destroyed of worlds

**Comment by GearTinkster** [comment]
i dont really like the idea of it just being dull executives so I'll think of another station trait to work on :)

**Comment by Rohesie** [comment]
> 
> 
> i dont really like the idea of it just being dull executives so I'll think of another station trait to work on :)

VIPs don't need to be dull executives.
You could, in fact, make the system generic and pick from specific VIP flavors, in which you could include something similar but more theme-fitting.
There's royalty in the modern days, including princesses, they are just not full medieval stereotypes.

But work on what you're interested, that's the healthiest way to contribute.

**Comment by EOBGames** [comment]
> > I fully agree with Potato here, and will be a little less diplomatic about it: the general idea is good, but the theme is not:
> 
> I am become loremaster, destroyed of worlds

I am become potato, crispy when fried

**Comment by Rohesie** [comment]
> I am become potato, crispy when fried

Hell. I blame you two joining the maintainer team not too far apart form each other, both being rather polite in your interactions, and thus forming a single entity in my mind.
